File: engine/dynamic/render_vulkan/src/setup/device.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2023, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/debug.hpp"
#include "argus/lowlevel/logging.hpp"
#include "argus/lowlevel/macros.hpp"

#include "internal/render_vulkan/module_render_vulkan.hpp"
#include "internal/render_vulkan/setup/device.hpp"
#include "internal/render_vulkan/setup/queues.hpp"
#include "internal/render_vulkan/setup/swapchain.hpp"

#include "vulkan/vulkan.h"

#include <algorithm>
#include <map>
#include <optional>
#include <utility>
#include <vector>

#include <cstdint>

namespace argus {
    static const uint32_t DISCRETE_GPU_RATING_BONUS = 10000;

    static std::optional<QueueFamilyIndices> _get_queue_family_indices(VkPhysicalDevice device,
            VkSurfaceKHR surface, const std::vector<VkQueueFamilyProperties> &queue_families) {
        QueueFamilyIndices indices = {};
        std::optional<uint32_t> transfer_queue;

        uint32_t i = 0;
        for (auto queue_family : queue_families) {
            if (queue_family.queueFlags & VK_QUEUE_TRANSFER_BIT && !(queue_family.queueFlags & VK_QUEUE_GRAPHICS_BIT)) {
                transfer_queue = i;
            }

            // check if queue family is suitable for graphics
            if (queue_family.queueFlags & VK_QUEUE_GRAPHICS_BIT) {
                indices.graphics_family = i;
            }

            VkBool32 present_support = false;
            vkGetPhysicalDeviceSurfaceSupportKHR(device, i, surface, &present_support);
            if (present_support) {
                indices.present_family = i;
            }

            i++;
        }

        indices.transfer_family = transfer_queue.value_or(indices.graphics_family);

        return indices;
    }

    static bool _is_device_suitable(VkPhysicalDevice device, VkSurfaceKHR probe_surface) {
        uint32_t ext_count;
        vkEnumerateDeviceExtensionProperties(device, nullptr, &ext_count, nullptr);

        std::vector<VkExtensionProperties> avail_exts(ext_count);
        vkEnumerateDeviceExtensionProperties(device, nullptr,
                &ext_count, avail_exts.data());

        std::set<std::string> req_exts(g_engine_device_extensions.cbegin(), g_engine_device_extensions.cend());

        for (const auto &ext : avail_exts) {
            req_exts.erase(ext.extensionName);
        }

        if (!req_exts.empty()) {
            Logger::default_logger().debug("Physical device '%s' is not suitable (missing required extensions)");
            return false;
        }

        auto sc_support = query_swapchain_support(device, probe_surface);
        if (sc_support.formats.empty()) {
            Logger::default_logger().debug("Physical device '%s' is not suitable (no available swap chain formats)");
            return false;
        }

        if (sc_support.present_modes.empty()) {
            Logger::default_logger().debug("Physical device '%s' is not suitable "
                                           "(no available swap chain present modes)");
            return false;
        }

        return true;
    }

    static uint32_t _rate_physical_device(VkInstance instance, VkPhysicalDevice device,
            std::vector<VkQueueFamilyProperties> queue_families) {
        UNUSED(instance);
        UNUSED(queue_families);

        uint32_t score = 0;

        VkPhysicalDeviceProperties props;
        vkGetPhysicalDeviceProperties(device, &props);

        VkPhysicalDeviceFeatures features;
        vkGetPhysicalDeviceFeatures(device, &features);

        if (props.deviceType == VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU) {
            score += DISCRETE_GPU_RATING_BONUS;
        }

        score += props.limits.maxImageDimension2D;

        //TODO: do some more checks at some point

        return score;
    }

    static std::pair<VkPhysicalDevice, QueueFamilyIndices> _select_physical_device(VkInstance instance,
            VkSurfaceKHR probe_surface) {
        uint32_t dev_count = 0;
        auto enum_res = vkEnumeratePhysicalDevices(instance, &dev_count, nullptr);
        if (enum_res) {
            Logger::default_logger().fatal("vkEnumeratePhysicalDevices returned error code %d", enum_res);
        }

        affirm_precond(dev_count > 0, "No physical video devices found");

        std::vector<VkPhysicalDevice> devs(dev_count);
        enum_res = vkEnumeratePhysicalDevices(instance, &dev_count, devs.data());

        VkPhysicalDevice best_dev = nullptr;
        QueueFamilyIndices best_dev_indices;
        uint32_t best_rating = 0;

        for (auto dev : devs) {
            VkPhysicalDeviceProperties dev_props;
            vkGetPhysicalDeviceProperties(dev, &dev_props);

            Logger::default_logger().debug("Considering physical device '%s'", dev_props.deviceName);

            uint32_t qf_props_count;
            vkGetPhysicalDeviceQueueFamilyProperties(dev, &qf_props_count, nullptr);
            if (qf_props_count == 0) {
                Logger::default_logger().debug("Physical device '%s' has no queue families", dev_props.deviceName);
                continue;
            }

            std::vector<VkQueueFamilyProperties> queue_families(qf_props_count);
            vkGetPhysicalDeviceQueueFamilyProperties(dev, &qf_props_count, queue_families.data());

            auto indices = _get_queue_family_indices(dev, probe_surface, queue_families);
            if (!indices.has_value()) {
                Logger::default_logger().debug("Physical device '%s' is not suitable", dev_props.deviceName);
                continue;
            }

            if (!_is_device_suitable(dev, probe_surface)) {
                continue;
            }

            auto rating = _rate_physical_device(instance, dev, queue_families);
            Logger::default_logger().debug("Physical device '%s' was assigned rating of %d", dev_props.deviceName,
                    rating);
            if (rating > best_rating) {
                best_dev = dev;
                best_dev_indices = indices.value();
                best_rating = rating;
            }
        }

        if (best_rating == 0) {
            Logger::default_logger().fatal("Failed to find suitable video device");
        }

        return std::make_pair(best_dev, best_dev_indices);
    }

    std::optional<LogicalDevice> create_vk_device(VkInstance instance, VkSurfaceKHR probe_surface) {
        VkPhysicalDevice phys_dev;
        QueueFamilyIndices qf_indices{};
        std::tie(phys_dev, qf_indices) = _select_physical_device(instance, probe_surface);
        VkPhysicalDeviceProperties phys_dev_props;
        vkGetPhysicalDeviceProperties(phys_dev, &phys_dev_props);

        Logger::default_logger().info("Selected video device %s", phys_dev_props.deviceName);

        std::set<uint32_t> unique_queue_families = { qf_indices.graphics_family, qf_indices.present_family,
                                                     qf_indices.transfer_family };
        std::vector<VkDeviceQueueCreateInfo> queue_create_infos;
        for (auto queue_id : unique_queue_families) {
            VkDeviceQueueCreateInfo queue_create_info{};
            queue_create_info.sType = VK_STRUCTURE_TYPE_DEVICE_QUEUE_CREATE_INFO;
            queue_create_info.queueFamilyIndex = queue_id;
            queue_create_info.queueCount = 1;
            auto queue_priority = 1.0f;
            queue_create_info.pQueuePriorities = &queue_priority;

            queue_create_infos.push_back(queue_create_info);
        }

        VkPhysicalDeviceFeatures dev_features{};

        VkDeviceCreateInfo dev_create_info{};
        dev_create_info.sType = VK_STRUCTURE_TYPE_DEVICE_CREATE_INFO;
        dev_create_info.pQueueCreateInfos = queue_create_infos.data();
        dev_create_info.queueCreateInfoCount = static_cast<uint32_t>(queue_create_infos.size());
        dev_create_info.pEnabledFeatures = &dev_features;
        dev_create_info.enabledExtensionCount = static_cast<uint32_t>(g_engine_device_extensions.size());
        dev_create_info.ppEnabledExtensionNames = g_engine_device_extensions.data();

        uint32_t layers_count;
        const char *const *layers;

        #ifdef _ARGUS_DEBUG_MODE
        layers_count = uint32_t(g_engine_layers.size());
        layers = g_engine_layers.data();
        #else
        layers_count = 0;
        layers = nullptr;
        #endif

        dev_create_info.enabledLayerCount = layers_count;
        dev_create_info.ppEnabledLayerNames = layers;

        VkDevice dev;
        auto rc = vkCreateDevice(phys_dev, &dev_create_info, nullptr, &dev);
        if (rc != VK_SUCCESS) {
            Logger::default_logger().fatal("Failed to create logical Vulkan device (rc: %d)", rc);
        }

        Logger::default_logger().debug("Successfully created logical Vulkan device");

        QueueFamilies queues{};
        vkGetDeviceQueue(dev, qf_indices.graphics_family, 0, &queues.graphics_family);
        vkGetDeviceQueue(dev, qf_indices.present_family, 0, &queues.present_family);
        vkGetDeviceQueue(dev, qf_indices.transfer_family, 0, &queues.transfer_family);

        auto *queue_mutexes = new QueueMutexes();

        return std::make_optional(LogicalDevice { phys_dev, dev, qf_indices, queues, queue_mutexes });
    }

    void destroy_vk_device(LogicalDevice device) {
        delete device.queue_mutexes;
        vkDestroyDevice(device.logical_device, nullptr);
    }
}

File: engine/modules/render_vulkan/src/setup/device.rs
use std::collections::HashSet;
use std::ffi::CStr;
use std::result::Result;
use ash::{khr, vk};
use argus_logging::{debug, info, warn};
use crate::setup::instance::VulkanInstance;
use crate::setup::LOGGER;
use crate::setup::queues::*;
use crate::setup::swapchain::query_swapchain_support;

const DISCRETE_GPU_RATING_BONUS: u64 = 10000;

pub(crate) const ENGINE_DEVICE_EXTENSIONS: &[&CStr] = &[
    khr::swapchain::NAME,
    khr::maintenance1::NAME,
];

#[derive(Clone)]
pub(crate) struct VulkanDevice {
    pub(crate) physical_device: vk::PhysicalDevice,
    pub(crate) logical_device: ash::Device,
    pub(crate) ext_khr_swapchain: khr::swapchain::Device,
    pub(crate) queue_indices: QueueFamilyIndices,
    pub(crate) queues: QueueFamilies,
    pub(crate) queue_mutexes: QueueMutexes,
    pub(crate) limits: vk::PhysicalDeviceLimits,
}

impl VulkanDevice {
    pub(crate) fn new(
        instance: &VulkanInstance,
        probe_surface: vk::SurfaceKHR
    ) -> Result<VulkanDevice, String> {
        let (phys_device, qf_indices) = select_physical_device(instance, probe_surface)
            .map_err(|err| err.to_string())?;

        let phys_dev_props = unsafe {
            instance.get_underlying().get_physical_device_properties(phys_device)
        };

        info!(
            LOGGER,
            "Selected video device {}",
            phys_dev_props.device_name_as_c_str().map(|s| s.to_string_lossy().to_string())
                .unwrap_or_else(|_| "(unknown)".to_owned()),
        );

        let unique_queue_families = HashSet::from([
            qf_indices.graphics_family,
            qf_indices.present_family,
            qf_indices.transfer_family,
        ]);
        let mut queue_create_infos = Vec::with_capacity(unique_queue_families.len());
        for queue_id in unique_queue_families {
            let queue_create_info = vk::DeviceQueueCreateInfo::default()
                .queue_family_index(queue_id)
                .queue_priorities(&[1.0]);

            queue_create_infos.push(queue_create_info);
        }

        let dev_features = vk::PhysicalDeviceFeatures::default()
            .independent_blend(true);

        let ext_names: Vec<*const i8> = ENGINE_DEVICE_EXTENSIONS.iter()
            .map(|ext| ext.as_ptr())
            .collect();

        let dev_create_info = vk::DeviceCreateInfo::default()
            .queue_create_infos(queue_create_infos.as_slice())
            .enabled_features(&dev_features)
            .enabled_extension_names(ext_names.as_slice());

        let device = unsafe {
            instance.get_underlying().create_device(phys_device, &dev_create_info, None)
                .map_err(|err| err.to_string())?
        };

        debug!(LOGGER, "Successfully created logical Vulkan device");

        let sc_device = khr::swapchain::Device::new(instance.get_underlying(), &device);

        let graphics_queue = unsafe { device.get_device_queue(qf_indices.graphics_family, 0) };
        let transfer_queue = unsafe { device.get_device_queue(qf_indices.transfer_family, 0) };
        let present_queue = unsafe { device.get_device_queue(qf_indices.present_family, 0) };

        let phys_device_limits = unsafe {
            let props = instance.get_underlying().get_physical_device_properties(phys_device);
            props.limits
        };

        Ok(VulkanDevice {
            physical_device: phys_device,
            logical_device: device,
            ext_khr_swapchain: sc_device,
            queue_indices: qf_indices,
            queues: QueueFamilies {
                graphics_family: graphics_queue,
                transfer_family: transfer_queue,
                present_family: present_queue,
            },
            queue_mutexes: Default::default(),
            limits: phys_device_limits,
        })
    }

    pub(crate) fn khr_swapchain(&self) -> &khr::swapchain::Device {
        &self.ext_khr_swapchain
    }

    pub(crate) fn destroy(self) {
        unsafe { self.logical_device.destroy_device(None); }
    }
}

fn get_queue_family_indices(
    device: vk::PhysicalDevice,
    surface: vk::SurfaceKHR,
    surface_instance: &khr::surface::Instance,
    queue_families: &Vec<vk::QueueFamilyProperties>,
) -> Result<QueueFamilyIndices, String> {
    let mut graphics_family: Option<u32> = None;
    let mut transfer_family: Option<u32> = None;
    let mut present_family: Option<u32> = None;

    let mut i: u32 = 0;
    for queue_family in queue_families {
        if transfer_family.is_none() &&
            queue_family.queue_flags.contains(vk::QueueFlags::TRANSFER) &&
            !queue_family.queue_flags.contains(vk::QueueFlags::GRAPHICS) {
            transfer_family = Some(i);
        }

        // check if queue family is suitable for graphics
        if graphics_family.is_none() &&
            queue_family.queue_flags.contains(vk::QueueFlags::GRAPHICS) {
            graphics_family = Some(i);
        }

        let device_has_present_support = unsafe {
            surface_instance.get_physical_device_surface_support(device, i, surface)
                .map_err(|err| err.to_string())?
        };
        if device_has_present_support {
            present_family = Some(i);
        }

        i += 1;
    }

    let Some(graphics_family) = graphics_family else {
        return Err("Device does not support any queue families with VK_QUEUE_VIDEO_BIT".to_owned());
    };

    let Some(present_family) = present_family else {
        return Err(
            "Device does not support any queue families that can present to a surface".to_owned()
        );
    };

    let indices = QueueFamilyIndices {
        graphics_family,
        transfer_family: transfer_family.unwrap_or(graphics_family),
        present_family,
    };

    Ok(indices)
}

fn is_device_suitable(
    instance: &VulkanInstance,
    device: vk::PhysicalDevice,
    probe_surface: vk::SurfaceKHR,
) -> bool {
    let enum_ext_res = unsafe {
        instance.get_underlying().enumerate_device_extension_properties(device)
    };
    let available_exts = match enum_ext_res {
        Ok(exts) => exts,
        Err(err) => {
            warn!(LOGGER, "Failed to enumerate device extension properties: {:?}", err);
            return false;
        }
    };
    let available_ext_names: Vec<_> = available_exts.iter()
        .filter_map(|ext| {
            match ext.extension_name_as_c_str() {
                Ok(name) => Some(name),
                Err(_) => {
                    warn!(
                        LOGGER,
                        "Encountered invalid (corrupted?) Vulkan extension name: {:?}",
                        ext.extension_name,
                    );
                    return None;
                }
            }
        })
        .collect();

    let required_exts = ENGINE_DEVICE_EXTENSIONS;

    for ext in required_exts {
        if !available_ext_names.iter().any(|name| name == ext) {
            debug!(
                LOGGER,
                "Physical device '{:?}' is not suitable (missing required extensions)",
                device,
            );
            return false;
        }
    }

    let sc_support = match unsafe { query_swapchain_support(instance, device, probe_surface) } {
        Ok(support) => support,
        Err(err) => {
            warn!(
                LOGGER,
                "Unable to query swapchain support for Vulkan device: {}",
                err.to_string(),
            );
            return false;
        }
    };
    if sc_support.formats.is_empty() {
        debug!(
            LOGGER,
            "Physical device '{:?}' is not suitable (no available swapchain formats)",
            device,
        );
        return false;
    }

    if sc_support.present_modes.is_empty() {
        debug!(
            LOGGER,
            "Physical device '{:?}' is not suitable (no available swapchain present modes)",
            device,
        );
        return false;
    }

    true
}

fn rate_physical_device(
    instance: &VulkanInstance,
    device: vk::PhysicalDevice,
    _queue_families: &Vec<vk::QueueFamilyProperties>,
) -> Option<u64> {
    let mut score: u64 = 0;

    // SAFETY: In principle this is safe so long as the underlying function in
    // the Vulkan driver is sound.
    let props = unsafe { instance.get_underlying().get_physical_device_properties(device) };
    let features = unsafe { instance.get_underlying().get_physical_device_features(device) };

    if features.independent_blend == 0 {
        warn!(LOGGER, "Cannot use device {:?} which does not support independentBlend", device);
        return None;
    }

    if props.device_type == vk::PhysicalDeviceType::DISCRETE_GPU {
        score += DISCRETE_GPU_RATING_BONUS;
    }

    score += props.limits.max_image_dimension2_d as u64;

    //TODO: do some more checks at some point

    Some(score)
}

fn select_physical_device(
    instance: &VulkanInstance,
    probe_surface: vk::SurfaceKHR,
) -> Result<(vk::PhysicalDevice, QueueFamilyIndices), String> {
    let devices = unsafe {
        instance.get_underlying().enumerate_physical_devices()
            .map_err(|err| err.to_string())?
    };

    if devices.is_empty() {
        return Err("No physical video devices found".to_owned());
    }

    let mut best_device: Option<(vk::PhysicalDevice, QueueFamilyIndices)> = None;
    let mut best_rating: u64 = 0;

    for device in devices {
        // SAFETY: In principle this is safe so long as the underlying function in
        // the Vulkan driver is sound.
        let dev_props = unsafe {
            instance.get_underlying().get_physical_device_properties(device)
        };
        let dev_name = dev_props.device_name_as_c_str()
            .map_err(|err| err.to_string())?
            .to_string_lossy();

        debug!(LOGGER, "Considering physical device '{}'", dev_name);

        // SAFETY: In principle this is safe so long as the underlying function in
        // the Vulkan driver is sound.
        let queue_families = unsafe {
            instance.get_underlying().get_physical_device_queue_family_properties(device)
        };

        let Ok(indices) = get_queue_family_indices(
            device,
            probe_surface,
            instance.khr_surface(),
            &queue_families,
        ) else {
            debug!(LOGGER, "Device {} is not suitable (unable to probe queue families)", dev_name);
            continue;
        };

        if !is_device_suitable(instance, device, probe_surface) {
            continue;
        }

        let Some(rating) = rate_physical_device(instance, device, &queue_families) else {
            continue;
        };

        debug!(LOGGER, "Physical device '{}' was assigned rating of {}", dev_name, rating);
        if rating > best_rating {
            best_device = Some((device, indices));
            best_rating = rating;
        }
    }

    match best_device {
        Some(dev) => Ok(dev),
        None => Err("Failed to find suitable video device".to_owned()),
    }
}
